src/platform/datapath_winuser.c

```diff
@@ -2780,47 +2770,48 @@ CxPlatSocketContextUninitialize(
     CxPlatSocketContextRelease(SocketProc);
 }
 
-void*
+CXPLAT_POOL_OBJECT*
 RioRecvBufferAllocate(
     _In_ uint32_t Size,
     _In_ uint32_t Tag,
     _Inout_ CXPLAT_POOL* Pool
     )
 {
-    CXPLAT_DATAPATH_PARTITION* DatapathProc =
-        CXPLAT_CONTAINING_RECORD(Pool, CXPLAT_DATAPATH_PARTITION, RioRecvPool);
-    CXPLAT_DATAPATH* Datapath = DatapathProc->Datapath;
+    CXPLAT_POOL_OBJECT* Object = CxPlatLargeAlloc(Size, Tag);
```

Comment:
Checking I understand correctly, we need this workaround with the CXPLAT_POOL_OBJECT because we have both a special memory allocator (`VirtualAlloc`) and some "on construction" logic to register the buffer that has been combined with the allocation?

It reminds me of C++ `new` keyword that basically call in succession `operator new` then the constructor. It could make sense to modify the pool to accept an "allocator" and a "constructor", to abstract some of that logic

Reply:
Yeah, it's kind of like that. But I don't want to split it up any more than it is. These are already very special case scenarios that can afford handling more stuff.